<!DOCTYPE html><html lang="en"><head><script async src="https://www.googletagmanager.com/gtag/js?id=UA-114897551-4"></script><script>window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'UA-114897551-4');
</script><script type="text/javascript" src="//platform-api.sharethis.com/js/sharethis.js#property=5ac2443d1fff98001395ab6c&amp;product=sticky-share-buttons" async="async"></script><title>Condition Systems in an Exceptional Language - Chris Houser | Coder Coacher - Coaching Coders</title><meta content="Condition Systems in an Exceptional Language - Chris Houser - All technical stuff in one place" name="description"><meta name="keywords" content="education, coding, programming, technology, nodejs, mongodb, software, computer science, engineering, teaching, coaching, coder, learning, java, kotlin, machine learning, AI, ML, tech talks, angular, javascript, js, typescript"><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="stylesheet" href="/css/font.css"><link rel="stylesheet" href="/css/bootstrap.css"><link rel="stylesheet" href="/css/style.css"><link rel="stylesheet" href="/css/coder-coacher.css"></head><body><div class="container-fluid"><h1 class="site-title"><a href="/">Coder Coacher</a></h1><hr><h4 class="site-subtitle text-right">Coaching Coders</h4></div><div id="amzn-assoc-ad-99d6751e-2392-4004-ad16-73aa8385d9d0"></div><script async src="//z-na.amazon-adsystem.com/widgets/onejs?MarketPlace=US&amp;adInstanceId=99d6751e-2392-4004-ad16-73aa8385d9d0"></script><div class="post__breadcrumb"><div class="container"><ol class="breadcrumb"><li><a href="/">Coder Coacher</a></li><li><a href="/ClojureTV/">ClojureTV</a></li><li class="active">â¤µ</li></ol></div></div><h2 class="post__title"><b>Condition Systems in an Exceptional Language - Chris Houser</b></h2><h5 class="post__date">2015-11-17</h5><div class="container"><div class="video-responsive"><iframe width="560" height="315" src="https://www.youtube.com/embed/zp0OEDcAro0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe></div><div class="post__text">good morning everybody how are you we
having fun my name take does say Chris
Houser but you can all call me ciao sir
I'm gonna do this alright so and I'm
gonna be talking about condition systems
if I can get this thing to work all
right this is not good all right we're
gonna do it this way okay here's my
outline I'm going to talk about a
problem and I'm going to go into detail
and exaggerate if necessary to make it
sound really bad then I'm going to
describe a solution and it'll be a great
solution you could even call it a silver
bullet
now I'm gonna after that I'm gonna
introduce a library that implements that
solution and I'm gonna give it a clever
name and then talk about only the good
things in the library and because it's
me I'm probably never actually gonna
release the library so all right that's
not that's not my outline my actual
outline is I'm gonna talk about some of
the options you have when it comes to
handling errors and then we're gonna
talk about common lisp it's condition
system provides a lot of those options
for you so you can choose am i speaking
odd enough can everybody hear me in the
back corners okay okay and then I'm
gonna look at some we're gonna show
briefly some of the libraries that are
already available for closure that
implement condition systems of their own
and then we're gonna talk about a sort
of a built in condition system that's
already in closure that you can use if
you need it without a library and then
I'm gonna talk about white none of them
are silver bullets but they all may have
be good options that you'd want to use
in various circumstances okay this this
is the happy path it wanders through
beautiful and clever libraries it's
where we like to spend our time and it
is the setting for most conference talks
it is a place where memory doesn't run
out networks are never disrupted
directories are never missing and discs
never fill up
a lovely place to be and not where we're
going to be spending the next 40 minutes
this is the sad path this is where rites
fail because the disk is full data can't
flow because the pipe is closed
connections fail because the network is
down constructors air out because memory
is full and people constantly try to
divide things by zero so what do you do
when you find yourself on the sad path
so here is a here's a solution and
answer from the 70s the action taken
after detecting a software error should
be uniform for all components in the
system this leads to the difficult
question of what action to take when an
error is detected the best action is
immediately to terminate the program
okay but we've heard this advice before
right where have we heard the let it
crash Erlang also interestingly tru64
unix which was a older UNIX operating
system if I remember correctly even a
single nullpointerexception write the D
referencing of a null pointer would lead
to the entire kernel the entire
operating system crashing and producing
stack traces that might even worry a
closure programmer
so this is an option that that that
makes sense and as it turns out tru64 is
incredibly stable because the
engineering team was motivated to
prevent this from happening right but
what about apps with interactive user
interfaces mobile apps web browsers and
that sort of thing you don't really want
them crashing anytime there's even the
slightest unexpected behavior right so
we've got a range of options here and
I'm gonna talk briefly about each of
them but you see at the top here sort of
the more dramatic harder to ignore
responses and down at the bottom are
easier to ignore or more subtle
responses and this probably isn't a
complete list but it should give you
sort of the sense of the range of things
you can do and the extremes are both a
little scary right crashing every time
might be worrisome but also just
recovering and and going forward might
cause problems too if you needed to know
about that error so let's just look at
each of these briefly because it turns
out that all across the spectrum
somebody has chosen this as a default
for their for their system so entering a
debugger is what common list does by
default when you signal an error you
don't catch it or anything mmm throwing
an exception to something of course
we're all familiar with special return
values so this is kind of a broad
category I think Java and here we can
see some closure code if you if you take
the square root of a negative number you
get back the value and a n right not a
number so that's a special value
indicating that something went wrong
closure takes this and use it in a lot
of places if you take the first item of
an empty collection you'll get nil
indicating that it wasn't able to answer
your question and I think maybes that
maybe monads nothing value falls in this
category as well it's a return value
that indicates something went wrong
setting an error flag right sees error
number is a little bit like this you set
a flag off to the side somewhere that
you have to check later to find out
whether an error happened and actually
we use this in closure closures agents
have a couple of different error modes
and one of them is to set a flag inside
the agent itself to indicate that
something went wrong and you can use
agent errors to
go back and check and see if there is an
error sitting there waiting for you and
finally recover and keep going
so another UNIX operating system you may
have heard of I believe I understand
correctly if you dereference a null
pointer inside the Linux kernel you'll
get an oops which is a little message in
a log that you probably never look at so
I think it's kind of interesting that we
have UNIX operating systems that have
chosen opposite ends of the spectrum
here so these are all potentially
reasonable choices now I'm gonna scoot
that off to the side here a little bit
and mention three other things about who
chooses the response that's going to be
taken so I'm quite confident that
they're they've got to be more options
here but these are the ones I was able
to find the detection sites the moment
you discover there's an error you could
choose one of these responses the
enclosing dynamic scope sometimes can
influence the option that ends up being
chosen
and finally the the object itself will
sometimes contain an error handler and
so anything that can create or mutate
that object may be able to choose one of
these options so this is probably one of
the more practical slides in this talk
so one of my main points is think about
this anytime that you are especially if
you're writing a library or something
where you notice that you have an error
and you need to propagate it to someone
think about your options there's a wide
range of options don't just assume that
throwing an exception is the right thing
to do think about what who's going to
want to be able to control it and what
kinds of options you'd like to have them
be able to choose between we'll talk a
little bit more about about how you can
implement any of those options to talk
about it we're going to use some
examples from Peter Siebels book
practical common lisp
trying to get all the different
combinations of possible options is hard
and I think he did a very good job of
coming up with an approachable example
so we're gonna go through that in a
little bit of detail okay so we're gonna
look at four functions just for this
example and then we're gonna change them
a little bit over time as we look at
different topics so this is the
lowest-level function in our stack parse
log entry and it's job is to check to
see if the input text which is gonna be
one line of text to see whether or not
it's well-formed and if it is we're
going to go ahead and parse it
successfully in this example just
returning a value because that's the
happy path we don't care about today for
the sad path we're gonna throw an
exception so you can there see we're
using X info which is a closure
exception constructor and giving it some
parameters about the error and then
throwing it this is gonna be called from
inside a function called parse log file
and the purpose of this function is to
open the file and split into lines for
processing and this provides a enclosing
dynamic scope for parse log entry right
so the call there to parse log entry
means that everything inside the body of
the lower function is happening within
the dynamic scope of the upper function
and I want to distinguish that from
lexical scope in case anybody's
unfamiliar the lexical or syntactic
scope is just what's physically up there
so for example the let form defines a
lines local which is only available to
that that keep line that's calling it so
you can't refer to lines down inside the
the lower function it's in the left it's
in the dynamic scope but not the lexical
okay we're gonna keep zooming out here
here's the next function up analyze log
the job of this one is to take a single
log file and go through all the entries
and analyze each of them more happy path
that we're not going to look a whole lot
at and then this top level function this
is gonna for our example is going to
represent the entire rest of the
application
above this it's responsible for finding
all the log files and parsing and
analyzing each so this is the highest
level dynamic scope and all the rest are
are contained inside it in a dynamic way
so we can go back to this slide and see
some of the choices that have been made
right at the detection site when we
noticed there was a problem we threw an
exception so that's one set of options
that you can choose but having thrown an
exception it doesn't end right there
right that that exception bubbles up
through the stack of dynamic scope and
when it hits the top if you were to run
that program straight from Java from the
command line or whatever the whole
process would crash but crashing here is
a bit of a problem because it means that
a single malformed log entry right in
one of those files that it looks through
if any one of them is malformed the
whole process stops right there all the
remaining log files and all the
remaining log entries aren't even going
to be looked at so this is why we have
try-catch right we want to be able to
catch those exceptions and prevent them
from ruining the entire process so
here's one way that we can change this
function so this is at the sort of in
the middle level here where we've got
this keep function that's representing a
loop right this is where we're going
through each of the items and we can put
this try-catch around the parse log
entry that was there before and now if
it's a malformed log entry we can return
nil and that nil sort of communicates
back to the keep that what it's supposed
to do at this point is to skip this
entry right and go on to the next one
after it so from our set of options here
it looks like this where now the
enclosing dynamic scope has chosen to
recover from the error and keep going by
skipping the entry so sorry let's see
yeah okay so here are the two important
functions that we just modified right
the parse log entry and the parse log
file and so now we're starting to see a
separation of concerns where we are
choosing or discovering that there is an
error in the lower level function at the
top of the screen we've identified the
problem but then down in parse log file
that defines how to respond right and
particularly how to skip an entry and
it's important that this be where we
define how to skip an entry because that
nil and the keep are communicating with
each other essentially write that nil
means skip an entry but this parse log
file is also still doing all of its file
parsing stuff and is also the place
where we're choosing to skip the entry
right so it knows how to skip the entry
but it's also the place where the policy
of choosing to do it is is implemented
so I would like to separate that a
little bit more and give the control to
somebody else to decide whether or not
to skip the entry and this is where
try-catch starts to peter out in its
ability to do that for you so we're
gonna switch to common list condition
system except instead of using common
lisp itself we're going to use a pair of
libraries called slingshot and swell
slingshot is a library by Steven Gilardi
and huben Duncan wrote a library and
topic called swell that together provide
an interface that's very similar to
common lists condition system they're
not always called condition systems
sometimes called resumable exceptions or
reasonable conditions or resumable
errors or just restarts so those are the
set of feature names that are sometimes
used to talk about these capabilities
we're about to look at and my point here
of showing the common lisp names is not
that I expect everybody here is familiar
with those but to show that there's a
very close correspondence between what
slingshot chose to do and and the common
lisp behaviors
okay so now we're gonna switch our
example over to using slingshot and so
you can see the first thing we have to
do is that the that lowest level
function parse log entry we're gonna
replace closures built in throw with
slingshots throw Plus now you can see
the parameters there are essentially the
same a little bit different order but
it's doing roughly the same kind of
thing okay but now at parse log file you
can see that we are no longer choosing
whether or not to skip instead we're
naming an entry called skip log entry
and that's a that's actually a restart
so we're naming a restart called skip
log entry and encoding how to skip which
is just a return a nil there right so
now at the parse log file level all
we're doing is providing the ability to
skip but we haven't chosen to do it yet
so now at a higher level up in a log
analyzer and of course again this this
function represents anything higher in
the dynamic stack so any thing that
calls down into down into analyze log
could choose to do this and we've got
this handler bind that we're wrapping
around the entire body of this loop
slingshot uses a predicate up there so
that that hash paren function at the top
is is a predicate saying that if this is
the kind of entry of error that has been
signaled then what should we do and the
choice of what to do is invoke the
restart named Skip log entry so now
we've separated concerns a little more
we've got a higher level function
choosing what to do in a lower level
function implementing how to do it
but now that we've got named restarts
why should we stop at one right so one
option is to skip log entry but we could
do other things as well now if you
remember a skip blog entry needed to be
where it was because there was a loop
there that was what we were skipping
it's an iterative thing and you want to
skip one iteration but there's other
things we can do at the lower level so
down in parse log entry again at the
lowest level here we're providing a
couple other restarts and these I
imagine are probably some of the most
common ones that people might implement
one is use value
and so this means that some higher-level
code can choose to invoke that and
provide a different output for this
parse log entry function so something at
a high level can say if you have any
kind of error instead of skipping it or
throwing an exception here is a value
I'd like you to put in the output in its
place maybe even more interesting is
reparse entry where a higher level code
can provide a function that takes the
text that was read out of the file and
change it around maybe with I don't know
erectus or if there's some common typo
or something that would fix the problem
it can fix it up and then notice it just
calls right back into parse log entry
with that new fixed up text okay so does
that making sense to everybody who kind
of where we are seeing some nodding
heads that's good so I just want to
point out that that we already have a
ton of options here with just this one
ability to separate these restarts from
the from the detection code and this is
I think where the bulk of the power
comes from so here again we have these
options so now we've talked about both
the detection site and the enclosing
dynamic scope and how those interact oh
you have a question
that's right well that's right so it
effectively I mean it's I mean you can
see the that's right instead of the
throw plus right so it'll come down
there and instead and when that throw
plus happens if it has been a sign from
the higher dynamic scope it will instead
go into the reparse entry and and call
parse log entry again so then it will
not revisit them that's right yep
so those are just named functions that
are available for use by those invoke
restarts that we saw at the higher level
thank you it's good question okay and
this is something you can't do with just
plain exceptions right so right so we've
talked about these I have a brief segue
here for this last one the object
creator or mutator it doesn't really fit
in very smoothly with the rest of the
conditions system you could make it be
so you could choose to use them but I
just want to make sure that we at least
look at one a short example of how it
can work I don't want to go into any
detail except to show that this is a
feature of agents again so this is the
other error mode for agents you can
provide an error handler that will have
some side effect when an error is
detected in this case a joke about
Swiper no swiping
okay so here are some of the conditions
system libraries that are already
available for closure
the first one is error kits it was
created in 2009 which is about three
months before closure 1.0 came out and I
think it's a testament to closure
stability that in order to get it to
work for this talk I had to change two
lines having to do with dynamic scope
and then everything else just worked so
that's was that six year old code that I
was working fine slingshot and swell
we've already seen rival is won by Chris
Jane who and it uses pattern matching
instead of pattern matching on the
objects that are raised instead of just
an inheritance hierarchy matching which
is what trycatch usually does right
bwo conditions is by Ben Wolfson and
this one was interesting because it's
based on a library for the ML language
which is another language that like
closure doesn't claim to have a
condition system so how is this even
possible that we've got all these
libraries on closure and mentioned here
of a library from ml that all provide
condition systems on languages that
don't claim to have them themselves well
here's an answer to that question from
what 97 so this paper is primarily about
formalizing dynamic binding but then it
goes on to show some of the uses of it
and one of them is when combined with an
abort operator you can divine define
these resumable exceptions which that
Greek stuff there at the bottom defines
and with help from Jim Dewey a co-worker
I can say with some confidence that it
really does although I'm not sure I
could explain it to you again now so an
abort is like throw an exception and the
idea of abort was formalized in a 1987
paper which is authored by four people
including dan Friedman who has spoken at
a couple of previous closure conferences
so this is sort of the this is the
underlying idea and I'm not sure that
this is the first time this idea was
ever had but it was formalized here and
this is the idea that all of the
libraries that provide conditions
systems in closure are built on top of
it's the same concept okay so we've got
lots of options so why don't we just use
one of these libraries and maybe you
should but there is a potential problem
and I think it can be called a
composition problem so let's assume for
a moment that you're you've started
writing some awesome app right and you
start using a couple of great libraries
now these libraries of course have
dependencies of their own so let's
suppose they each use a condition system
library
now I don't I'm not picking on these two
in particular my point here is that
there are different ones right so what
happens now in awesome app if you start
trying to handle these errors and
provide restarts and that sort of thing
now you've got a problem right you've
got different kinds of restarts each of
these libraries have their own syntax
their own macros for handling these and
although I believe there's it's probably
not undefined the behavior so it may not
be technically the hard kind of
composition problem it's certainly gonna
make your code trickier than then you
would want it to be and so what actually
happens is most closure library authors
know this and they choose not to burden
their applications with this kind of
problem so so now we have all these
features that we've just talked about
that could be useful in some situations
and yet we're not allowed to use the
libraries that provide them for fear of
causing problems for applications above
us so what should we do well the last
chapter of joy of closure has a
mentioned focus starts us on the path of
doing this ourselves and it turns out
you can get most of the features of a
condition system just with built-in
features of closure without any library
helping out so that's what we're gonna
go look at here for the next few slides
but it starts like this the basic idea
is you define a function representing an
error so in this case the malformed log
entry error but you make sure that you
flag it as a dynamic function dynamic
var so now instead of throwing an
exception you just call that function
alright so that one change that one
adjustment of pulling that throw out
into a separate function marking that
function is dynamic that right there
gives you a ton of control and is
exactly what focus points out in the
last chapter of joy closure so now that
you've done that here at the top level
in log analyzer you can see that we've
replaced the handler bind the special
handler bind macro with just closures
own built-in binding form where we are
binding to that that error function a
new behavior right we're providing
new behavior instead of throwing
exception you just return constantly nil
and we know from having looked at the
rest of the program that nil is gonna go
into the keep and the keep is gonna skip
the entry so right there
we've decided to skip the entry all
without using any separate special
libraries and I guess it's worth
pointing out here that binding the
dynamic scope the binding sets up is
exactly the same kind of scope the
try-catch does it's the same same
boundaries which is also the same
behavior as common lisps handle bind and
is also exactly what Moroz paper shows
okay
so here's all the code that you need to
change to switch from using slingshot
and swell to using the built-in
behaviors of closure but we've again
we've committed the same problem the
same error that we did before in that
actually before I mentioned that I want
to point out that this is it this is I
guess I've already kind of talked about
that but this is all you need to do
right this one little change pulling
that function that that throw out into
its own function gives you a lot of
flexibility but one of the problems we
have at this point is that we is one
similar to one we saw earlier which is
that this function down here log
analyzer is now used exploiting its
knowledge of the of the interior
functions right it knows that there's a
keep and that nil is gonna somehow skip
an entry so we don't really want to
expose the implementation of one
function into the behavior of another
right so so we want to do that ability
to name some restarts right so here's
how you can do that so we already have
that that error as a function but now we
can define each of the restarts as
another dynamic each of them as their
own dynamic function and at this point
if you were to call one of them you'd
immediately get an error because there's
no behavior specified right that's also
exactly the same behavior as you get in
common list when you try to invoke a
restart that hasn't been defined but now
that we have those names we can provide
implementations for them so we're now
down back again at the bottom level
function parse log entry then you can
see we can now define both use value and
reparse entry just like we did before
use value is just going to return the
value that was provided
by the high-level function to replace
the bad entry and reparse entry will
call back into this this same function
so this is the same behavior we saw
before now just using closures binding
built-in form okay and then this is how
you use it so this is up at the top
level again representing anywhere above
this in the in the in the program you
can bind that malformed log entry error
function so now instead of building in
the knowledge of constantly nil we can
just call the function use value and
provide the value that we would like to
be used instead of the bad entry so at
this point we've caught up with all of
sciples examples from practical Common
Lisp so right there you can now go
through that chapter of the book and and
implement all of the features that he's
showing off about Common Lisp just using
closures built-in features it is worth
pointing out that the book also contains
a nice little example of how to use a
condition system to handle non error
situations right so now that you don't
always have to end up throwing exception
and causing something that looks like a
crash it might be worth thinking about
using in other circumstances he's got an
example of using it in a actually I
didn't write down I think it's a another
kind of parser where he's using it to
communicate between functions within the
parser but that means that my point to
you is when you are thinking about these
these condition systems you might want
to broaden your idea of what an error is
a little bit and realize the
similarities that it has to a callback
alright this is essentially now a
callback with a sort of dynamic scope
for when it's available so although this
is where the books examples stop Common
Lisp goes quite a bit further and we can
follow so what about how to catch oh I
see yes so a catch Clause can handle not
just a single kind of exception right
but a whole tree in of an inheritance
hierarchy so it turns out we can do
something similar as well so here's an
example of what you might want to do so
we have this malformed blog entry error
which is the only one we've been talking
about so far as if that was the one kind
of error you would ever have but if you
want you can say that that is like
or general malformed entry for things
that aren't logs I suppose and we've got
some math errors I made up over here to
sort of format an inheritance tree right
so if we were using exceptions this
would actually be Java inheritance but
if we're going to use functions to
represent the throwing of those errors
what we do is we just delegate so you
can see down here we've got malformed
blog entry error again and instead of
throwing exception it's calling the
function that it would like to defer to
or the error that it wants to defer to
and that defers on up until we finally
hit the general error function up at the
top which will just throw the exception
and so this allows users to bind
anywhere in that tree and catch all of
the all the errors below it in the in
the delegation tree so another feature
that common list provides for its error
handlers is that after one of them has
been invoked it can decline to handle it
so we can look at the inputs and say oh
I thought I knew just based on the the
error name or whatever I thought I knew
how I was gonna be able to handle this
but looking at the particular inputs
that are available I I declined to
handle this I want something above me
something that was bound previously to
do to handle that so this is a you know
a handler declining to handle which just
sounds like that to me
so here's how we can do this with our
with our built-in functions right so
this is getting a little bit Messier and
now you can start to see why people
actually wrote libraries for this right
but the basic behavior is you use a let
binding write a local lexical binding to
grab the old value the old behavior of
the error before you enter your dynamic
binding so now you've got this this
function this behavior that represents
you declining to handle the malformed
blog entry error and then down inside
your handler so here when the actual
malformed blog entry error shows up you
can look at the inputs and base them
choose on whether you want to do
something that was already to find one
of these restarts like use-value or you
can go ahead and call decline and
that'll trigger whatever behavior was
was defined at a larger scope okay
so we've looked at more and more complex
ways to handle these errors and more and
more code you can put around your your
error conditions so here's a quote I
really like largely because it's from
the 50s those who regularly code for
fast electronic computers will have
learned from bitter experience that a
large fraction of the time spent in
preparing calculations for the machine
is taking up taken up and removing the
blunders that have been made in drying
up the program don't you love the
language with the aid of common sense
and checking subroutines the majority of
mistakes are quickly found and rectified
some errors however are sufficiently
obscure to escape detection for a
surprisingly long time so the question
that I would like to ask at this point
is you know if we step back a moment and
think about this do we really want
complex code in obscures corners obscure
corners of our sad paths right we
already have on a sad path this is stuff
that doesn't happen that often and
probably doesn't get checked very
carefully so how many checking
subroutines are you going to have to
write to make sure that you don't have
errors in your complex error handling
and so as in all areas of design we need
to work to achieve a balance so
here are my thoughts first of all
consider alternatives to just throwing
an exception it may not always be the
right choice try out the built in
condition system just just try it every
time that you're about to write a throw
just don't write it there move it out
into a separate function make that
function dynamic and and try it if
nobody uses it
you haven't caused any particular harm
if you document it and people find it
useful you've actually provided a lot of
flexibility to the users of your library
and you can rest assured that you can
continue to extend that as necessary to
achieve really as much complexities as
you're likely to need and finally if you
aren't writing a library that's going to
be exposed to the wider public and
instead of writing an application where
you have some control of the code all
the way up to the highest levels you
might actually consider using a
condition system that's out there and
has maybe put then had a little more
thought put into it than whatever
binding code that you're writing as you
go along so just a parting thought for
my friend Karen Meier vast opportunity
the vast opportunities that we software
developers have to make models that
prove and expand human knowledge but
were too busy plumbing so I'm sorry to
have talked to you for a half-hour here
about nothing but plumbing but I think
she's right so think about this make
sure you do a good job consider how to
handle your errors but then you know
move on and do something actually
important and don't squander a lot of
your time thinking about condition
systems all right that's how I have</div></div><div class="container-fluid bottom-ad"><div id="amzn-assoc-ad-6a809dda-347a-4187-8a86-91faf94575da"></div><script async src="//z-na.amazon-adsystem.com/widgets/onejs?MarketPlace=US&amp;adInstanceId=6a809dda-347a-4187-8a86-91faf94575da"></script></div><div class="text-center">We are a participant in the Amazon Services LLC Associates Program, an affiliate advertising program designed to provide a means for us to earn fees by linking to Amazon.com and affiliated sites.</div><script>(function(w, d){
    var b = d.getElementsByTagName('body')[0];
    var s = d.createElement("script"); s.async = true;
    var v = !("IntersectionObserver" in w) ? "8.6.0" : "10.4.2";
    s.src = "https://cdnjs.cloudflare.com/ajax/libs/vanilla-lazyload/" + v + "/lazyload.min.js";
    w.lazyLoadOptions = {};
    b.appendChild(s);
}(window, document));</script></body></html>